File: walk.rs
#[derive(Clone)]
struct Node {
    name: String,
    kids: Vec<Node>,
    // parent: Option<&'a Node<'a>>,
}

impl Node {
    fn leaf(name: &str) -> Node {
        Node::parent(name, vec![])
    }

    fn parent(name: &str, kids: Vec<Node>) -> Node {
        Node {
            name: name.into(),
            kids,
            // parent: None,
        }
    }
}

// fn init_parents(tree: &mut Node) {
//     for kid in &mut tree.kids {
//         kid.parent = Some(&tree);
//         init_parents(kid);
//     }
// }

fn walk_depth<Action>(tree: &Node, action: &mut Action, depth: usize)
where
    Action: FnMut(&Node, usize),
{
    action(tree, depth);
    // for kid in &tree.kids {
    for i in 0..tree.kids.len() {
        let kid = &tree.kids[i];
        walk_depth(kid, action, depth + 1);
    }
}

fn walk<Action: FnMut(&Node, usize)>(tree: &Node, action: &mut Action) {
    walk_depth(tree, action, 0);
}

fn print_tree(tree: &Node) {
    walk(tree, &mut |node, depth| {
        println!("{:depth$}{name}", "", depth = 2 * depth, name = node.name);
    });
}

fn calc_total_depth(tree: &Node) -> usize {
    let mut total = 0;
    walk(tree, &mut |_node, depth| {
        total += depth;
    });
    total
}

fn process(intro: &Node) {
    // -> &Node {
    let tree = Node::parent(
        "root",
        vec![
            intro.clone(),
            Node::parent("one", vec![Node::leaf("two"), Node::leaf("three")]),
            Node::leaf("four"),
        ],
    );
    // init_parents(&mut tree);
    // let internal_intro = &tree.kids[0];
    // tree.kids.push(Node::leaf("outro"));
    // println!("{}", internal_intro.name);
    // internal_intro.parent = Some(&tree);
    print_tree(&tree);
    let mut total_depth = 0;
    for _ in 0..200_000 {
        total_depth += calc_total_depth(&tree);
    }
    println!("Total depth: {total_depth}");
    // &tree
    // internal_intro
}

fn main() {
    let intro = Node::leaf("intro");
    process(&intro);
}

File: walk.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const print = @import("std").debug.print;

const Node = struct {
    name: []const u8 = "",
    kids: std.ArrayList(Node),
    parent: ?*Node = null,

    fn leaf(allocator: Allocator, name: []const u8) Node {
        return parent(allocator, name, &.{}) catch unreachable;
    }

    fn parent(allocator: Allocator, name: []const u8, kids: []Node) !Node {
        var kids_list = std.ArrayList(Node).init(allocator);
        try kids_list.ensureTotalCapacityPrecise(kids.len);
        try kids_list.appendSlice(kids);
        return Node{ .name = name, .kids = kids_list };
    }

    fn deinit(self: Node) void {
        // TODO Reverse order to crash!
        for (self.kids.items) |*kid| {
            kid.deinit();
        }
        self.kids.deinit();
    }
};

fn initParents(tree: *Node) void {
    for (tree.kids.items) |*kid| {
        kid.parent = tree;
        initParents(kid);
    }
}

fn walkDepth(comptime Result: type, value: Result, tree: Node, comptime action: fn (Result, Node, usize) Result, depth: usize) Result {
    var result = action(value, tree, depth);
    // for (tree.kids.items) |kid| {
    var i = @as(usize, 0);
    while (i < tree.kids.items.len) : (i += 1) {
        const kid = tree.kids.items[i];
        result = walkDepth(Result, result, kid, action, depth + 1);
    }
    return result;
}

fn walk(comptime Result: type, value: Result, tree: Node, comptime action: fn (Result, Node, usize) Result) Result {
    return walkDepth(Result, value, tree, action, 0);
}

fn printNode(_: void, node: Node, depth: usize) void {
    var i = @as(usize, 0);
    while (i < 2 * depth) : (i += 1) {
        print(" ", .{});
    }
    print("{s}\n", .{node.name});
}

fn printTree(tree: Node) void {
    var v: void = undefined;
    walk(void, v, tree, printNode);
}

fn calcTotalDepthNode(total: usize, _: Node, depth: usize) usize {
    return total + depth;
}

fn calcTotalDepth(tree: Node) usize {
    return walk(usize, 0, tree, calcTotalDepthNode);
}

fn process(intro: *Node) !Node {
    const allocator = intro.kids.allocator;
    var tree = try Node.parent(allocator, "root", &.{
        intro.*,
        try Node.parent(allocator, "one", &.{
            Node.leaf(allocator, "two"),
            Node.leaf(allocator, "three"),
        }),
        Node.leaf(allocator, "four"),
    });
    // defer tree.deinit();
    initParents(&tree);
    const internal_intro = &tree.kids.items[0];
    // try tree.kids.append(Node.leaf(allocator, "outro"));
    print("{s}\n", .{internal_intro.name});
    printTree(tree);
    var total_depth = @as(usize, 0);
    var calc_repeats = @as(usize, 0);
    while (calc_repeats < 2) : (calc_repeats += 1) {
        total_depth += calcTotalDepth(tree);
    }
    print("Total depth: {}\n", .{total_depth});
    // return internal_intro;
    return tree;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    var intro = Node.leaf(allocator, "intro");
    const tree = try process(&intro);
    defer tree.deinit();
    _ = tree;
    print("{s}\n", .{tree.name});
}
